igvm/commands.py

```diff
@@ -235,45 +255,74 @@ def vm_build(vm_hostname, run_puppet=True, debug_puppet=False, postboot=None,
-                True,
-            ))
-            vm.dataset_obj['hypervisor'] = \
-                vm.hypervisor.dataset_obj['hostname']
+        if vm.dataset_obj['igvm_operation_mode'] == 'aws':
+            file_loader = FileSystemLoader('igvm/templates/')
+            env = Environment(loader=file_loader)
+            template = env.get_template('aws_user_data.cfg')
 
-        if vm.hypervisor.vm_defined(vm) and vm.is_running():
-            raise InvalidStateError(
-                '"{}" is still running.'.format(vm.fqdn)
+            user_data = template.render(
+                hostname=vm.dataset_obj['hostname'].rstrip('.ig.local'),
+                fqdn=vm.dataset_obj['hostname'],
+                puppet_master=vm.dataset_obj['puppet_master'],
+                puppet_ca=vm.dataset_obj['puppet_ca'],
             )
 
-        if rebuild and vm.hypervisor.vm_defined(vm):
-            vm.hypervisor.undefine_vm(vm)
+            vm.aws_build(
+                run_puppet=run_puppet,
+                debug_puppet=debug_puppet,
+                postboot=user_data
+            )
+        elif vm.dataset_obj['igvm_operation_mode'] == 'kvm':
+            if vm.hypervisor:
+                es.enter_context(_lock_hv(vm.hypervisor))
+            else:
+                vm.hypervisor = es.enter_context(_get_best_hypervisor(
+                    vm,
+                    ['online', 'online_reserved'] if allow_reserved_hv
+                    else ['online'],
+                    True,
+                ))
+                vm.dataset_obj['hypervisor'] = \
+                    vm.hypervisor.dataset_obj['hostname']
+
+            if vm.hypervisor.vm_defined(vm) and vm.is_running():
+                raise InvalidStateError(
+                    '"{}" is still running.'.format(vm.fqdn)
+                )
 
-        vm.build(
-            run_puppet=run_puppet,
-            debug_puppet=debug_puppet,
-            postboot=postboot,
-        )
+            if rebuild and vm.hypervisor.vm_defined(vm):
+                vm.hypervisor.undefine_vm(vm)
+
+            vm.build(
+                run_puppet=run_puppet,
+                debug_puppet=debug_puppet,
+                postboot=postboot,
+            )
+        else:
+            raise NotImplementedError(
+                'This operation is not yet supported for {}'.format(
+                    vm.dataset_obj['igvm_operation_mode']))
 
         vm.dataset_obj.commit()
 
 
-@with_fabric_settings   # NOQA: C901
+@with_fabric_settings  # NOQA: C901
 def vm_migrate(vm_hostname, hypervisor_hostname=None,
                run_puppet=False, debug_puppet=False,
                offline=False, offline_transport='drbd',
                allow_reserved_hv=False, no_shutdown=False):
     """Migrate a VM to a new hypervisor."""
+
     with ExitStack() as es:
         vm = es.enter_context(
             _get_vm(vm_hostname, allow_retired=True)
         )
+
+        if vm.dataset_obj['igvm_operation_mode'] != 'kvm':
```

Comment:
Thanks for your extensive review and feedback. We will leave it like is for now, since it's pretty easy to rename this, if we're planning to virtualize reasonable things without igvm. But I noted it